### PR TITLE
fix: disallow SSE-C encrypted objects on replicated buckets

### LINKED ISSUE #16463: Disallow SSE-C when using bucket replication (or site replication)
**Is your feature request related to a problem? Please describe.**

Currently SSE-C (Server-Side Encryption with Client-Managed Keys) is not compatible with Bucket Replication (as [documented](https://min.io/docs/minio/linux/administration/server-side-encryption/server-side-encryption-sse-c.html#sse-c-is-incompatible-with-bucket-replication)) and by implication also not compatible with server-side site replication (because as far as I can tell server-side site replication uses bucket replication under the hood).

The problem is that SSE-C is something out of control for the server i.e. the client can decide whether to use it or not. This leads to a situation where for data safety or availability reasons the MinIO server has been configured to use server-side site replication, a client unwittingly uses SSE-C and those objects are never replicated. This is even worse if one or multiple clients modify the same SSE-C object in several sites, leading to what _seems to be (from the client's point of view) a data inconsistency_, because the changes to the object are never replicated.

Problems:

* Clients receive no error message when trying to PUT an SSE-C object even if the server is using bucket replication
* The replication status reports nothing is wrong

Am I right in assuming the same problem applies to Object Lifecycle Management i.e. it does not work for objects encrypted via SSE-C?

**Describe the solution you'd like**

* Either support SSE-C in combination with Bucket Replication (preferred solution)
* or make client operations using SSE-C fail
* or reflect the fact that some objects can't be replicated (because they are SSE-C encrypted) in the replication status

Maybe worthwhile mentioning that AWS supports replicating SSE-C encrypted objects.

> S3 Replication supports objects that are encrypted with SSE-C.

[Source](https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-config-for-kms-objects.html#replicationSSEC)

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: disallow SSE-C encrypted objects on replicated buckets

## Motivation and Context
Reject SSE-C uploaded objects on replicated buckets.

fixes #16463

## How to test this PR?
On a replicated bucket you can try

```
./mc cp --encrypt-key="sitea/bucket=32byteslongsecretkeymustbegiven2" /etc/hosts sitea/bucket/sse-c-file.txt
```

Would fail with

```
mc: <ERROR> Failed to copy `/etc/hosts`. SSE-C encryption parameters are not supported on replicated bucket.
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: docs/bucket/replication/README.md [hunk 1]
```diff
@@ -264,6 +264,16 @@ When Bucket Versioning with excluded prefixes are configured objects matching th
 
 In the above sample config, objects under prefixes matching any of the `ExcludedPrefixes` glob patterns will neither be versioned nor replicated.
 
+### SSE-C Encryption
+
+MinIO does not support SSE-C encrypted objects on replicated buckets, any application uploading SSE-C encrypted objects will be rejected with an error on replicated buckets.
+
+**Rationale**
+
+- SSE-C requires application remember the keys for all GET/PUT operations, any unfortunate loss of keys would automatically mean the objects cannot be accessed anymore.
```
Comment (krisis) [inline docs/bucket/replication/README.md]:
```suggestion
- SSE-C requires application to remember the keys for all GET/PUT operations, any unfortunate loss of keys would automatically mean the objects cannot be accessed anymore.
```

**Review by krisis — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/encryption-v1.go [hunk 2]
```diff
@@ -55,7 +55,8 @@ var (
 	// Additional MinIO errors for SSE-C requests.
 	errObjectTampered = errors.New("The requested object was modified and may be compromised")
 	// error returned when invalid encryption parameters are specified
-	errInvalidEncryptionParameters = errors.New("The encryption parameters are not applicable to this object")
+	errInvalidEncryptionParameters     = errors.New("The encryption parameters are not applicable to this object")
+	errInvalidEncryptionParametersSSEC = errors.New("SSE-C encryption parameters are not supported on this bucket")
```
Comment (kannappanr) [inline cmd/encryption-v1.go:59]: Should we say not supported on a replicated bucket

Comment (harshavardhana) [reply cmd/encryption-v1.go:59]: That error returned to the client @kannappanr this is just a placeholder error.

Comment (harshavardhana) [reply cmd/encryption-v1.go:59]: > mc: <ERROR> Failed to copy `/etc/hosts`. SSE-C encryption parameters are not supported on replicated bucket.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**